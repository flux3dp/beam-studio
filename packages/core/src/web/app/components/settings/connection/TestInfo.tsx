import React, { useMemo } from 'react';
import TestState from 'app/constants/connection-test';
import useI18n from 'helpers/useI18n';

import { promarkModels } from 'app/actions/beambox/constant';
import styles from './TestInfo.module.scss';

interface Props {
  testState: TestState;
  connectionCountDown: number;
  firmwareVersion?: string;
}

const TestInfo = ({ testState, connectionCountDown, firmwareVersion = '' }: Props): JSX.Element => {
  const tConnect = useI18n().initialize.connect_machine_ip;
  const { model } = useMemo(() => {
    const queryString = window.location.hash.split('?')[1] || '';
    const urlParams = new URLSearchParams(queryString);

    return { model: urlParams.get('model') };
  }, []);

  const isPromark = promarkModels.has(model);

  const renderBasicConnectionInfo = (
    reached: string = tConnect.check_ip,
    unreachableError = tConnect.unreachable
  ) => {
    if (testState === TestState.NONE) {
      return null;
    }

    let status = 'OK ✅';

    switch (testState) {
      case TestState.IP_TESTING:
        status = '';
        break;
      case TestState.IP_FORMAT_ERROR:
        status = `${tConnect.invalid_ip}${tConnect.invalid_format}`;
        break;
      case TestState.IP_UNREACHABLE:
        status = unreachableError;
        break;
      default:
        break;
    }

    return <div id="ip-test-info" className={styles.info}>{`${reached}... ${status}`}</div>;
  };

  const renderConnectionTestInfo = () => {
    if (testState < TestState.CONNECTION_TESTING) return null;
    let status = 'OK ✅';
    if (testState === TestState.CONNECTION_TESTING) status = `${connectionCountDown}s`;
    else if (testState === TestState.CONNECTION_TEST_FAILED) status = 'Fail';
    return (
      <div id="machine-test-info" className={styles.info}>
        {`${tConnect.check_connection}... ${status}`}
      </div>
    );
  };

  const renderFirmwareVersion = () => {
    if (testState < TestState.CAMERA_TESTING) return null;
    return <div className={styles.info}>{`${tConnect.check_firmware}... ${firmwareVersion}`}</div>;
  };

  const renderCameraTesting = () => {
    if (testState < TestState.CAMERA_TESTING) return null;
    let status = '';
    if (testState === TestState.TEST_COMPLETED) status = 'OK ✅';
    else if (testState === TestState.CAMERA_TEST_FAILED) status = 'Fail';
    return <div className={styles.info}>{`${tConnect.check_camera}... ${status}`}</div>;
  };

  return (
    <div className={styles.container}>
      {isPromark
        ? renderBasicConnectionInfo(
            tConnect.check_swiftray_connection,
            tConnect.check_swiftray_connection_unreachable
          )
        : renderBasicConnectionInfo()}
      {renderConnectionTestInfo()}
      {renderFirmwareVersion()}
      {renderCameraTesting()}
      {testState === TestState.TEST_COMPLETED && (
        <div className={styles.info}>{tConnect.succeeded_message}</div>
      )}
    </div>
  );
};

export default TestInfo;
