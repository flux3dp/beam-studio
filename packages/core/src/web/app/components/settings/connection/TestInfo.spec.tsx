import React from 'react';
import { render } from '@testing-library/react';

import TestState from 'app/constants/connection-test';

import TestInfo from './TestInfo';

jest.mock('helpers/useI18n', () => () => ({
  initialize: {
    connect_machine_ip: {
      invalid_ip: 'invalid_ip',
      invalid_format: 'invalid_format',
      unreachable: 'unreachable',
      check_ip: 'check_ip',
      check_connection: 'check_connection',
      check_firmware: 'check_firmware',
      check_camera: 'check_camera',
      succeeded_message: 'succeeded_message',
    },
  },
}));

const STATES_TO_TEST = [
  { text: 'none', state: TestState.NONE },
  { text: 'ip format error', state: TestState.IP_FORMAT_ERROR },
  { text: 'ip testing', state: TestState.IP_TESTING },
  { text: 'ip unreachable', state: TestState.IP_UNREACHABLE },
  { text: 'connection testing', state: TestState.CONNECTION_TESTING },
  { text: 'camera testing', state: TestState.CAMERA_TESTING },
  { text: 'camera test failed', state: TestState.CAMERA_TEST_FAILED },
  { text: 'test completed', state: TestState.TEST_COMPLETED },
];

describe('test TestInfo', () => {
  it.each(STATES_TO_TEST)('should correctly when state is $text', ({ state }) => {
    const { container } = render(<TestInfo testState={state} firmwareVersion="4.0.0" connectionCountDown={10} />);
    expect(container).toMatchSnapshot();
  });
});
