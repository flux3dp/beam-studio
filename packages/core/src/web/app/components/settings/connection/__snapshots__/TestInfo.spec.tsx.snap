// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test TestInfo should correctly when state is camera test failed 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="info"
      id="ip-test-info"
    >
      check_ip... OK ✅
    </div>
    <div
      class="info"
      id="machine-test-info"
    >
      check_connection... OK ✅
    </div>
    <div
      class="info"
    >
      check_firmware... 4.0.0
    </div>
    <div
      class="info"
    >
      check_camera... Fail
    </div>
  </div>
</div>
`;

exports[`test TestInfo should correctly when state is camera testing 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="info"
      id="ip-test-info"
    >
      check_ip... OK ✅
    </div>
    <div
      class="info"
      id="machine-test-info"
    >
      check_connection... OK ✅
    </div>
    <div
      class="info"
    >
      check_firmware... 4.0.0
    </div>
    <div
      class="info"
    >
      check_camera... 
    </div>
  </div>
</div>
`;

exports[`test TestInfo should correctly when state is connection testing 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="info"
      id="ip-test-info"
    >
      check_ip... OK ✅
    </div>
    <div
      class="info"
      id="machine-test-info"
    >
      check_connection... 10s
    </div>
  </div>
</div>
`;

exports[`test TestInfo should correctly when state is ip format error 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="info"
      id="ip-test-info"
    >
      check_ip... invalid_ipinvalid_format
    </div>
  </div>
</div>
`;

exports[`test TestInfo should correctly when state is ip testing 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="info"
      id="ip-test-info"
    >
      check_ip... 
    </div>
  </div>
</div>
`;

exports[`test TestInfo should correctly when state is ip unreachable 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="info"
      id="ip-test-info"
    >
      check_ip... unreachable
    </div>
  </div>
</div>
`;

exports[`test TestInfo should correctly when state is none 1`] = `
<div>
  <div
    class="container"
  />
</div>
`;

exports[`test TestInfo should correctly when state is test completed 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="info"
      id="ip-test-info"
    >
      check_ip... OK ✅
    </div>
    <div
      class="info"
      id="machine-test-info"
    >
      check_connection... OK ✅
    </div>
    <div
      class="info"
    >
      check_firmware... 4.0.0
    </div>
    <div
      class="info"
    >
      check_camera... OK ✅
    </div>
    <div
      class="info"
    >
      succeeded_message
    </div>
  </div>
</div>
`;
